src/LanguageServer/Protocol/Handler/RelatedDocuments/RelatedDocumentsHandler.cs

```diff
@@ -0,0 +1,147 @@
+{
+    /// <summary>
+    /// Cache where we store the data produced by prior requests so that they can be returned if nothing of significance
+    /// changed. The version key is produced by combining the checksums for project options <see
+    /// cref="ProjectState.GetParseOptionsChecksum"/> and <see cref="DocumentStateChecksums.Text"/>
+    /// </summary>
+    private readonly VersionedPullCache<(Checksum parseOptionsChecksum, Checksum textChecksum)?> _versionedCache = new(nameof(RelatedDocumentsHandler));
+
+    public bool MutatesSolutionState => false;
+    public bool RequiresLSPSolution => true;
+
+    private static async Task<(Checksum parseOptionsChecksum, Checksum textChecksum)> ComputeChecksumsAsync(Document document, CancellationToken cancellationToken)
+    {
+        var project = document.Project;
+        var parseOptionsChecksum = project.State.GetParseOptionsChecksum();
+
+        var documentChecksumState = await document.State.GetStateChecksumsAsync(cancellationToken).ConfigureAwait(false);
+        var textChecksum = documentChecksumState.Text;
+
+        return (parseOptionsChecksum, textChecksum);
+    }
+
+    public TextDocumentIdentifier GetTextDocumentIdentifier(VSInternalRelatedDocumentParams requestParams)
+        => requestParams.TextDocument;
+
+    /// <summary>
+    /// Retrieve the previous results we reported.  Used so we can avoid resending data for unchanged files.
+    /// </summary>
+    private static ImmutableArray<PreviousPullResult>? GetPreviousResults(VSInternalRelatedDocumentParams requestParams)
+        => requestParams.PreviousResultId != null && requestParams.TextDocument != null
+            ? [new PreviousPullResult(requestParams.PreviousResultId, requestParams.TextDocument)]
+            // The client didn't provide us with a previous result to look for, so we can't lookup anything.
+            : null;
+
+    public async Task<VSInternalRelatedDocumentReport[]?> HandleRequestAsync(
+        VSInternalRelatedDocumentParams requestParams, RequestContext context, CancellationToken cancellationToken)
+    {
+        context.TraceInformation($"{this.GetType()} started getting related documents");
+
+        // The progress object we will stream reports to.
+        using var progress = BufferedProgress.Create(requestParams.PartialResultToken);
+
+        context.TraceInformation($"PreviousResultId={requestParams.PreviousResultId}");
+
+        var solution = context.Solution;
+        var document = context.Document;
+        Contract.ThrowIfNull(solution);
+        Contract.ThrowIfNull(document);
+
+        context.TraceInformation($"Processing: {document.FilePath}");
+
+        var relatedDocumentsService = document.GetLanguageService<IRelatedDocumentsService>();
+        if (relatedDocumentsService == null)
+        {
+            context.TraceInformation($"Ignoring document '{document.FilePath}' because it does not support related documents");
```

Comment:
could consider an early return for readability

Reply:
sure.

Reply:
done.